From fc849ba4c7a5517e5cd92e38999d7f7d1a074f7b Mon Sep 17 00:00:00 2001
From: Yiwen Jiang <jiangyiwen@huawei.com>
Date: Fri, 26 Nov 2021 16:26:58 +0800
Subject: [PATCH] kvm: arm/arm64: add irqsave for lpi_cache_lock
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: fc849ba4c7a5517e5cd92e38999d7f7d1a074f7b


euleros inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4J0W7
CVE: NA

-------------------------------------------------

lpi_cache_lock can be called in irq context,
so it should use irqsave spinlock.

Signed-off-by: Yiwen Jiang <jiangyiwen@huawei.com>
Reviewed-by: Hailiang Zhang <zhang.zhanghailiang@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Chaochao Xing <xingchaochao@huawei.com>
Reviewed-by: Xiangyou Xie <xiexiangyou@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kvm/vgic/vgic-its.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/kvm/vgic/vgic-its.c b/arch/arm64/kvm/vgic/vgic-its.c
index ec6d5b9fc57c..4f31880803d4 100644
--- a/arch/arm64/kvm/vgic/vgic-its.c
+++ b/arch/arm64/kvm/vgic/vgic-its.c
@@ -593,15 +593,16 @@ static struct vgic_irq *vgic_its_check_cache(struct kvm *kvm, phys_addr_t db,
 {
 	struct vgic_dist *dist = &kvm->arch.vgic;
 	struct vgic_irq *irq;
+	unsigned long flags;
 	int cpu;
 	int cacheid;
 
 	cpu = smp_processor_id();
 	cacheid = cpu % LPI_TRANS_CACHES_NUM;
 
-	raw_spin_lock(&dist->lpi_translation_cache[cacheid].lpi_cache_lock);
+	raw_spin_lock_irqsave(&dist->lpi_translation_cache[cacheid].lpi_cache_lock, flags);
 	irq = __vgic_its_check_cache(dist, db, devid, eventid, cacheid);
-	raw_spin_unlock(&dist->lpi_translation_cache[cacheid].lpi_cache_lock);
+	raw_spin_unlock_irqrestore(&dist->lpi_translation_cache[cacheid].lpi_cache_lock, flags);
 
 	return irq;
 }
-- 
2.26.2

